components/tikv_util/src/codec/bytes.rs

```diff
@@ -278,6 +278,152 @@ pub fn decode_bytes_in_place(data: &mut Vec<u8>, desc: bool) -> Result<()> {
+
+    // Check the last group first
+    let last_encoded_group = &encoded[encoded.len() - ENC_GROUP_SIZE - 1..];
+    let last_group_len = raw.len() % ENC_GROUP_SIZE;
+    let last_raw_group = &raw[raw.len() - last_group_len..];
+    let pad_len = ENC_GROUP_SIZE - last_group_len;
+    const PADDING_BUF: [u8; 8] = [!ENC_MARKER; 8];
+    if last_encoded_group[ENC_GROUP_SIZE] != (ENC_MARKER - pad_len as u8)
+        || &last_encoded_group[..last_group_len] != last_raw_group
+        || last_encoded_group[last_group_len..ENC_GROUP_SIZE] != PADDING_BUF[..pad_len]
+    {
+        return false;
+    }
+
+    let mut encoded_group_ptr = last_encoded_group.as_ptr();
+    let mut raw_group_ptr = last_raw_group.as_ptr();
+    while encoded_group_ptr > encoded.as_ptr() {
+        unsafe {
+            let pad_ptr = encoded_group_ptr.sub(1);
+            encoded_group_ptr = pad_ptr.sub(ENC_GROUP_SIZE);
+            raw_group_ptr = raw_group_ptr.sub(ENC_GROUP_SIZE);
+            if *pad_ptr != ENC_MARKER
+                || slice::from_raw_parts(encoded_group_ptr, ENC_GROUP_SIZE)
+                    != slice::from_raw_parts(raw_group_ptr, ENC_GROUP_SIZE)
+            {
+                return false;
+            }
+        }
+    }
+    true
+}
+
+/// Returns whether `encoded` is encoded from `raw` using descending memory-comparable format.
+/// Returns `false` if `encoded` is invalid.
+pub fn is_encoded_from_desc(encoded: &[u8], raw: &[u8]) -> bool {
+    let group_count = raw.len() / ENC_GROUP_SIZE + 1;
+    if encoded.len() != group_count * (ENC_GROUP_SIZE + 1) {
+        return false;
+    }
+
+    // Bytes are compared in reverse order because in real cases like TiDB, if two keys
+    // are different, the last a few bytes are more likely to be different.
+
+    // Check the last group first
+    let last_encoded_group = &encoded[encoded.len() - ENC_GROUP_SIZE - 1..];
+    let last_group_len = raw.len() % ENC_GROUP_SIZE;
+    let last_raw_group = &raw[raw.len() - last_group_len..];
+    let pad_len = ENC_GROUP_SIZE - last_group_len;
+    const PADDING_BUF: [u8; 8] = [ENC_MARKER; 8];
+    if last_encoded_group[ENC_GROUP_SIZE] != !(ENC_MARKER - pad_len as u8)
+        || last_encoded_group[..last_group_len]
+            .iter()
+            .zip(last_raw_group)
+            .any(|(&encoded, &raw)| encoded != !raw)
+        || last_encoded_group[last_group_len..ENC_GROUP_SIZE] != PADDING_BUF[..pad_len]
+    {
+        return false;
+    }
+
+    let mut encoded_group_ptr = last_encoded_group.as_ptr();
+    let mut raw_group_ptr = last_raw_group.as_ptr();
+    while encoded_group_ptr > encoded.as_ptr() {
+        unsafe {
+            let pad_ptr = encoded_group_ptr.sub(1);
+            encoded_group_ptr = pad_ptr.sub(ENC_GROUP_SIZE);
+            raw_group_ptr = raw_group_ptr.sub(ENC_GROUP_SIZE);
+            if *pad_ptr != !ENC_MARKER
+                || *encoded_group_ptr != !*raw_group_ptr
```

Comment:
But we need to flip the bits...

Reply:
How about comparing `LHS: u64` and `!RHS: u64`?

Reply:
Encoded bytes have bad memory align, I doubt we can compare them as `u64`.

Reply:
If you wanted to be really fancy, you could unroll the loop into 4s, put the bytes into 256bit SIMD registers and compare those (not sure how much you will be able to do in SIMD, might be worth trying if this function is really hot).

Reply:
@sticnarf Of course we can, we can simply do two unaligned memory read (`mov` is enough) and then compare.

Reply:
@breeswish Not that sure, but I remember we cannot do that with an ARM CPU...

Reply:
Oh, I find std::ptr::read_unaligned probably helps.

Reply:
"Not that sure, but I remember we cannot do that with an ARM CPU..." we can have different implementations for different targets. I think it is good to find something fast for modern x86 since that is our main target, and then if necessary we can add different impls for ARM, etc.

Reply:
You can use `#[target_arch = "foo"]` like so https://doc.rust-lang.org/reference/conditional-compilation.html#target_arch